### PR TITLE
pub upgrade --major-versions

### PR DESCRIPTION
**Pull request by jonasfj**
Merged master into https://github.com/dart-lang/pub/pull/2619 and change the name of the flag and a few other things

CC @walnutdust

### CODE REVIEW

**Comment by walnutdust** [comment]
@googlebot I consent.

File: lib/src/command/upgrade.dart [hunk 1]
```diff
@@ -53,73 +54,196 @@ class UpgradeCommand extends PubCommand {
...
+      dryRun: _dryRun,
+      precompile: _precompile,
+    );
 
     _showOfflineWarning();
   }
 
-  Future<void> _runUpgradeNullSafety() async {
+  /// Return names of packages to be upgraded, and throws [UsageException] if
+  /// any indirect dependencies are given.
```
Comment (sigurdm) [inline lib/src/command/upgrade.dart:112]:
```suggestion
  /// any package names not in the direct dependencies or dev_dependencies are given.
```

File: lib/src/command/upgrade.dart [hunk 2]
```diff
@@ -53,73 +54,196 @@ class UpgradeCommand extends PubCommand {
...
       ...entrypoint.root.pubspec.devDependencies.keys
     ];
-    final upgradeOnly = argResults.rest.isEmpty ? directDeps : argResults.rest;
+    final toUpgrade = argResults.rest.isEmpty ? directDeps : argResults.rest;
 
     // Check that all package names in upgradeOnly are direct-dependencies
-    if (upgradeOnly.any((name) => !directDeps.contains(name))) {
-      final notDirectDeps =
-          upgradeOnly.where((name) => !directDeps.contains(name)).toList();
+    final indirectDeps = toUpgrade.where((n) => !directDeps.contains(n));
```
Comment (sigurdm) [inline lib/src/command/upgrade.dart:125]:
```suggestion
    final notInDeps = toUpgrade.where((n) => !directDeps.contains(n));
```

File: lib/src/command/upgrade.dart [hunk 3]
```diff
@@ -53,73 +54,196 @@ class UpgradeCommand extends PubCommand {
...
+  Future<void> _runUpgradeMajorVersions() async {
+    final toUpgrade = _directDependenciesToUpgrade();
+
+    final resolvablePubspec = stripVersionUpperBounds(
+      entrypoint.root.pubspec,
+      stripOnly: toUpgrade,
+    );
+
+    /// Solve [resolvablePubspec] in-memory and consolidate the resolved
+    /// versions of the packages into a map for quick searching.
```
Comment (sigurdm) [inline lib/src/command/upgrade.dart:155]: Nit:
```suggestion
    // Solve [resolvablePubspec] in-memory and consolidate the resolved
    // versions of the packages into a map for quick searching.
```

File: lib/src/command/upgrade.dart [hunk 4]
```diff
@@ -53,73 +54,196 @@ class UpgradeCommand extends PubCommand {
...
+        cache,
+        Package.inMemory(resolvablePubspec),
+      );
+    }, condition: _shouldShowSpinner);
+    for (final resolvedPackage in solveResult?.packages ?? []) {
+      resolvedPackages[resolvedPackage.name] = resolvedPackage;
+    }
+
+    /// Changes to be made to `pubspec.yaml`.
+    /// Mapping from original to changed value.
```
Comment (sigurdm) [inline lib/src/command/upgrade.dart:169]: Nit:
```suggestion
    // Changes to be made to `pubspec.yaml`.
    // Mapping from original to changed value.
```

**Review by sigurdm — APPROVED** [review]
LGTM with a few suggestions

File: lib/src/command/upgrade.dart [hunk 5]
```diff
@@ -150,15 +274,29 @@ be direct 'dependencies' or 'dev_dependencies', following packages are not:
...
+      // TODO(jonasfj): Stop abusing Entrypoint.global for dry-run output
+      await Entrypoint.global(
+        Package.inMemory(nullsafetyPubspec),
+        entrypoint.lockFile,
+        cache,
+        solveResult: solveResult,
+      ).acquireDependencies(
+        SolveType.UPGRADE,
+        dryRun: true,
+        precompile: _precompile,
```
[the remark this answers is not in the document]
Comment (jonasfj) [reply lib/src/command/upgrade.dart:289]: it's of no effect, I'm thinking that if in the future we start showing some precompilation information in dry-run mode that would be cool, but really it doesn't matter.

File: lib/src/command/upgrade.dart [hunk 6]
```diff
@@ -181,6 +181,11 @@ be direct 'dependencies' or 'dev_dependencies', following packages are not:
         continue;
       }
 
+      // If there is a dependency_override for dependency then we ship it.
```
Comment (sigurdm) [inline lib/src/command/upgrade.dart:184]:
```suggestion
    // Skip [dependency] If it has a dependency_override.
```

File: lib/src/command/upgrade.dart [hunk 7]
```diff
@@ -220,6 +225,18 @@ be direct 'dependencies' or 'dev_dependencies', following packages are not:
 
     _outputChangeSummary(changes);
 
+    // If any of the packages to upgrade are dependency overrides, then we
+    // show a warning.
```
Comment (sigurdm) [inline lib/src/command/upgrade.dart:229]:
```suggestion
    // Show a warning if any of the packages to upgrade have dependency overrides.
```

File: lib/src/command/upgrade.dart [hunk 8]
```diff
@@ -220,6 +225,18 @@ be direct 'dependencies' or 'dev_dependencies', following packages are not:
...
+    final toUpgradeOverrides = toUpgrade
+        .where(entrypoint.root.dependencyOverrides.containsKey)
+        .toList();
+    if (toUpgradeOverrides.isNotEmpty) {
```
Comment (sigurdm) [inline lib/src/command/upgrade.dart:233]:
```suggestion
    final hasUpgradeOverrides = toUpgrade
        .any(entrypoint.root.dependencyOverrides.containsKey);
    if (hasUpgradeOverrides) {
```

**Review by sigurdm — APPROVED** [review]
Only a few nits

File: lib/src/command/upgrade.dart (hunk 8 above)
Comment (jonasfj) [reply lib/src/command/upgrade.dart:233]: Aren't you just making it less useful by now showing which packages are affected?

Comment (sigurdm) [reply lib/src/command/upgrade.dart:233]: Sorry this was wrong. Disregard

**Pull request merged**